![pandas](http://pandas.pydata.org/_static/pandas_logo.png "Pandas Logo")


- Started by Wes MacKinney with a first release in 2011.
- Based on NumPy, it is the most used library for all things data.
- Motivated by the toolbox in R for manipulating data easily.
- A lot of names in Pandas come from R world.
- It is Open source (BSD)

https://pandas.pydata.org/

# Pandas 

```python
import pandas as pd
```

"*Pandas provides high-performance, easy-to-use data structures 
and data analysis tools in Python*"

- Self-describing data structures
- Data loaders to/from common file formats
- Plotting functions
- Basic statistical tools.


In [1]:
%reset -f
%matplotlib inline
%config InlineBackend.figure_format = 'retina'
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()
pd.set_option("display.max_rows", 8)
plt.rcParams['figure.figsize'] = (9, 6)

# [Series](https://pandas.pydata.org/pandas-docs/stable/dsintro.html#series)

- A Series contains a one-dimensional array of data, *and* an associated sequence of labels called the *index*.
- The index can contain numeric, string, or date/time values.
- When the index is a time value, the series is a [time series](https://en.wikipedia.org/wiki/Time_series).
- The index must be the same length as the data.
- If no index is supplied it is automatically generated as `range(len(data))`.

In [2]:
pd.Series([1,3,5,np.nan,6,8])

0    1.0
1    3.0
2    5.0
3    NaN
4    6.0
5    8.0
dtype: float64

In [3]:
pd.Series(index=pd.period_range('09/11/2017', '09/18/2017', freq="D"))

2017-09-11   NaN
2017-09-12   NaN
2017-09-13   NaN
2017-09-14   NaN
2017-09-15   NaN
2017-09-16   NaN
2017-09-17   NaN
2017-09-18   NaN
Freq: D, dtype: float64

### Exercise 13.1
- Create a text with `lorem` and count word occurences with a `collection.Counter`. Put the result in a `dict`.

### Exercise 13.2
- From the results create a Pandas series name latin_series with words in alphabetical order as index.

### Exercise 13.3

- Plot the series using 'bar' kind.

### Exercise 13.4
- Pandas provides explicit functions for indexing `loc` and `iloc`.
    - Use `loc` to display the number of occurrences of 'dolore'.
    - Use `iloc` to diplay the number of occurrences of the last word in index.

### Exercise 13.5
- Sort words by number of occurrences.
- Plot the Series.

## Full globe temperature between 1901 and 2000.

We read the text file and load the results in a pandas dataframe. 
In cells below you need to clean the data and convert the dataframe to a time series.

In [182]:
import os
here = os.getcwd()

filename = os.path.join(here,"data","monthly.land.90S.90N.df_1901-2000mean.dat.txt")

df = pd.read_table(filename, sep="\s+", 
                   names=["year", "month", "mean temp"])
df

year  month  mean temp
0     1880      1    -0.0235
1     1880      2    -0.4936
2     1880      3    -0.6785
3     1880      4    -0.2829
...    ...    ...        ...
1580  2011      9  -999.0000
1581  2011     10  -999.0000
1582  2011     11  -999.0000
1583  2011     12  -999.0000

[1584 rows x 3 columns]

### Exercise 13.6
- Insert a third column with value one named "day" with `.insert`.
- convert df index to datetime with `pd.to_datetime` function.
- convert df to Series containing only "mean temp" column.
- Display the beginning of the file with `.head`.
- Display the end of the file with `.tail`.
- In the dataset, -999.00 was used to indicate that there was no value for that year. Display values equal to -999 with `.values`. 
- Replace the missing value (-999.000) by `np.nan` 
- Once they have been converted to np.nan, missing values can be removed (dropped).
Remove missing values with `.dropna`.
- Generate a basic visualization using `.plot`.
- Convert df index from timestamp to period is more meaningfull since it was measured and averaged over the month. Use `to_period` method.

## Resampling

Series can be resample, downsample or upsample.
- Frequencies can be specified as strings: "us", "ms", "S", "T", "H", "D", "B", "W", "M", "A", "3min", "2h20", ...
- More aliases at http://pandas.pydata.org/pandas-docs/stable/timeseries.html#offset-aliases

### Exercise

- With `resample` method, convert df Series to 10 year blocks:

In [194]:
df.resample('10A').mean()


1880   -0.386485
1890   -0.316798
1900   -0.256431
1910   -0.247673
          ...   
1980    0.188519
1990    0.463572
2000    0.785452
2010    0.884700
Freq: 10A-DEC, Name: mean temp, Length: 14, dtype: float64

# Saving Work

[HDF5](https://support.hdfgroup.org/HDF5/) is widely used and one of the most powerful file format to store binary data. It allows to store both Series and DataFrames.

In [197]:
with pd.HDFStore("data/pandas_series.h5") as writer:
    df.to_hdf(writer, "/temperatures/full_globe")

# Reloading data

In [199]:
with pd.HDFStore("data/pandas_series.h5") as store:
    df = store["/temperatures/full_globe"]